### PR TITLE
Implement SAKE

### PR DESCRIPTION
**Pull request by ArnNag**
## Description
Implement the SAKE model

## Status
- [ ] Ready to go

### CODE REVIEW

File: devtools/conda-envs/test_env.yaml [hunk 1]
```diff
@@ -32,7 +32,11 @@ dependencies:
   - qcelemental=0.25.1
   - qcportal>=0.50
   - requests
+  - jax
+  - flax
```
Comment (wiederm) [inline devtools/conda-envs/test_env.yaml:36]: can we avoid `jax` and `flax` in this implementation?

File: devtools/conda-envs/test_env.yaml [hunk 2]
```diff
@@ -32,7 +32,11 @@ dependencies:
...
+  - versioneer
 
   # pip installs
   - pip:
       - schnetpack>=2.0.0
+      - git+https://github.com/ArnNag/sake.git@update_versioneer
```
Comment (wiederm) [inline devtools/conda-envs/test_env.yaml:42]: I am not sure what this installation is doing?

File: modelforge/potential/sake.py [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+    References:
+        https://openreview.net/pdf?id=3DIpIf3wQMC
+
+    """
+
+    def __init__(
+            self,
+            embedding_module: nn.Module,
+            nr_interaction_blocks: int,
+            nr_heads: int,
```
Comment (wiederm) [inline modelforge/potential/sake.py:24]: can you please also document this parameter?

File: modelforge/potential/sake.py [hunk 4]
```diff
@@ -0,0 +1,382 @@
...
+            torch.cat(
+                [h_ij_cat, h_ij_filtered, d_ij.unsqueeze(-1)],
+                dim=1
+            )
+        )
+
+        return h_ij_out
+
+
+class ExpNormalSmearing(nn.Module):
```
Comment (wiederm) [inline modelforge/potential/sake.py:346]: Can you please add docstrings explaining the functionality and use case, and the parameters that are passed?

File: modelforge/potential/sake.py [hunk 5]
```diff
@@ -0,0 +1,382 @@
...
+            radial_basis_module=radial_basis,
+            cutoff_module=cutoff,
+            activation=activation,
+        )
+        self.loss_function = loss
+        self.optimizer = optimizer
+        self.learning_rate = lr
+
+
+class ContinuousFilterConvolutionWithConcatenation(nn.Module):
```
Comment (wiederm) [inline modelforge/potential/sake.py:321]: Can you please add docstrings explaining the functionality and use case, and the parameters that are passed?

File: modelforge/potential/sake.py [hunk 6]
```diff
@@ -0,0 +1,382 @@
...
+        combinations_norm_square = (combinations_mean ** 2).sum(dim=-1)
+        return self.post_norm_mlp(combinations_norm_square)
+
+    def aggregate(self, h_ij_semantic, idx_i, nr_atoms):
+        expanded_idx_i = idx_i.view(-1, 1).expand_as(h_ij_semantic)
+        out_shape = (nr_atoms, self.nr_heads * self.nr_edge_basis)
+        zeros = torch.zeros(out_shape, dtype=h_ij_semantic.dtype, device=h_ij_semantic.device)
+        return zeros.scatter_add(0, expanded_idx_i, h_ij_semantic)
+
+    def get_semantic_attention(self, h_ij_edge, idx_i, d_ij, nr_atoms):
```
Comment (wiederm) [inline modelforge/potential/sake.py:240]: This is pretty dense code. I think comments and docstrings would help the reader to grasp what is going on here better.

File: modelforge/tests/test_sake.py [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+
+    pairlist = torch.cartesian_prod(torch.arange(nr_atoms), torch.arange(nr_atoms))
+    nr_pairs = nr_atoms ** 2
+    pairlist = pairlist.T
+    idx_i, idx_j = pairlist
+
+    key = jax.random.PRNGKey(1884)
+    h_e_att = jax.random.normal(key, (nr_atoms, nr_atoms, hidden_features * nr_heads))
+    x_minus_xt = jax.random.normal(key, (nr_atoms, nr_atoms, geometry_basis))
+    x_minus_xt_norm = jnp.linalg.norm(x_minus_xt, axis=-1, keepdims=True)
```
Comment (wiederm) [inline modelforge/tests/test_sake.py:208]: I think we can express this with pure numpy as well?

**Review by wiederm — COMMENTED** [review]
Thank you @ArnNag ! That looks great. I think the code would benefit from a few more comments and docstrings (I left comments in the most obvious paces). After that is addressed I will give the code a more in depth review.

File: modelforge/tests/test_sake.py (hunk 7 above)
Comment (ArnNag) [reply modelforge/tests/test_sake.py:208]: ah yes sorry i was in the middle of converting this

File: devtools/conda-envs/test_env.yaml (hunk 1 above)
Comment (ArnNag) [reply devtools/conda-envs/test_env.yaml:36]: I was implementing correctness tests against the original, should we remove this from the testing suite?

File: devtools/conda-envs/test_env.yaml (hunk 2 above)
Comment (ArnNag) [reply devtools/conda-envs/test_env.yaml:42]: This is also for the correctness tests

File: modelforge/tests/test_sake.py (hunk 7 above)
Comment (ArnNag) [reply modelforge/tests/test_sake.py:208]: This is for the correctness tests (see above)

File: modelforge/potential/sake.py [hunk 8]
```diff
@@ -0,0 +1,510 @@
...
+
+from .models import BaseNNP, LightningModuleMixin
+from .postprocessing import PostprocessingPipeline, NoPostprocess
+from .utils import Dense, scatter_softmax
+import torch
+import torch.nn.functional as F
+
+
+class ExpNormalSmearing(torch.nn.Module):
+    def __init__(self, cutoff_lower=0.0, cutoff_upper=5.0, num_rbf=50, trainable=True):
```
Comment (wiederm) [inline modelforge/potential/sake.py:14]: For the future: we should merge this with the RadialSymmetryFunction class in ANI and SchNET for a common set of Smearing classes.

File: modelforge/tests/test_sake.py [hunk 9]
```diff
@@ -0,0 +1,410 @@
...
+
+    mf_h, mf_x, mf_v = mf_sake_block(h, x, v, pairlist)
+
+    ref_h, ref_x, ref_v = ref_sake_interaction.apply(variables, h_jax, x_jax, v_jax, mask)
+
+    ref_h_is_nan = torch.from_numpy(onp.isnan(ref_h))
+    ref_x_is_nan = torch.from_numpy(onp.isnan(ref_x))
+    ref_v_is_nan = torch.from_numpy(onp.isnan(ref_v))
+
+    assert torch.allclose(torch.nan_to_num(mf_h, nan=0.0) * ~ref_h_is_nan,
```
Comment (wiederm) [inline modelforge/tests/test_sake.py:273]: locally, this test is failing on my machine.
The first row of the tensors that are compared:
```
tensor([[ 2.3430, -0.3628,  3.4058,  0.1804,  1.3299,  0.0351,  1.8090,  1.0771,
          1.1417,  0.7593,  0.0136],
tensor([[ 2.4133, -0.6323,  3.3515,  0.0421,  1.0885,  0.2227,  2.0435,  1.2833,
          0.9287,  0.6991, -0.1271],
```

**Pull request merged**